platform-dom/src/main/java/com/softicar/platform/dom/elements/input/auto/DomAutoCompleteDefaultInputEngine.java

```diff
@@ -0,0 +1,208 @@
+
+		this.cache = new DerivedObject<>(Cache::new);
+		this.loader = Collections::emptyList;
+		this.displayFunction = new DefaultDisplayFunction();
+		this.comparator = new DefaultComparator();
+	}
+
+	/**
+	 * Defines the {@link Supplier} for the elements.
+	 * <p>
+	 * This {@link Supplier} should prefetch all necessary data so that calls to
+	 * {@link #getDisplayString} for the returned elements are efficient.
+	 * <p>
+	 * No deduplication of elements is performed, since this is deemed to be the
+	 * obligation of the {@link Supplier}.
+	 *
+	 * @param loader
+	 *            the element {@link Supplier} (never <i>null</i>)
+	 * @return this
+	 */
+	public DomAutoCompleteDefaultInputEngine<T> setLoader(Supplier<Collection<T>> loader) {
+
+		this.loader = Objects.requireNonNull(loader);
+		return this;
+	}
+
+	/**
+	 * Adds the given {@link Objects} as a dependency to the internal
+	 * {@link DerivedObject}, which is used to cache the loaded elements.
+	 *
+	 * @param object
+	 *            the source object (never <i>null</i>)
+	 * @return this
+	 */
+	public DomAutoCompleteDefaultInputEngine<T> addDependsOn(Object object) {
+
+		cache.addDependsOn(object);
+		return this;
+	}
+
+	/**
+	 * Overrides the default implementation to determine the
+	 * {@link IDisplayString} of an element.
+	 * <p>
+	 * The default implementation relies on the elements implementing the
+	 * {@link IDisplayable} interface.
+	 *
+	 * @param displayFunction
+	 *            the display function (never <i>null</i>)
+	 * @return this
+	 */
+	public DomAutoCompleteDefaultInputEngine<T> setDisplayFunction(Function<T, IDisplayString> displayFunction) {
+
+		this.displayFunction = Objects.requireNonNull(displayFunction);
+		return this;
+	}
+
+	/**
+	 * Overrides the {@link Comparator} to order elements with equal
+	 * {@link IDisplayString} instances.
+	 * <p>
+	 * The default implementation relies on the elements implementing the
+	 * {@link Comparable} interface.
+	 *
+	 * @param comparator
+	 *            the {@link Comparator} (never <i>null</i>)
+	 * @return this
+	 */
+	public DomAutoCompleteDefaultInputEngine<T> setComparator(Comparator<T> comparator) {
+
+		this.comparator = Objects.requireNonNull(comparator);
+		return this;
+	}
+
+	@Override
+	public final IDisplayString getDisplayString(T element) {
+
+		return cache.get().getDisplayString(element);
+	}
+
+	@Override
+	public final Collection<T> findMatches(String pattern, int limit) {
+
+		return cache.get().findMatches(pattern, limit);
+	}
+
+	@Override
+	public void refresh() {
+
+		cache.invalidate();
+	}
+
+	private class Cache {
+
+		private final Map<String, T> stringToElementMap;
+		private final Map<T, String> elementToStringMap;
+
+		public Cache() {
+
+			this.stringToElementMap = new DomAutoCompleteDisplayStringDeduplicator<>(displayFunction, comparator).apply(loader.get());
+			this.elementToStringMap = stringToElementMap.entrySet().stream().collect(Collectors.toMap(Entry::getValue, Entry::getKey));
```

Comment:
Using `Collectors.toMap` might be dangerous here. Quoting from its Javadoc:
> There are no guarantees on the type, [...] of the {@code Map} returned.

So, an Exception could occur if, for example, `T` is non-`Comparable` (it's unbounded, after all) and a `TreeMap` is created.

Reply:
- unlikely, i think that will break lots of code if they ever change that to TreeMap
- anyway, now doing it manually, so sad